### PR TITLE
Alee/reserve

### PR DESCRIPTION
**Pull request by zander-xyz**
- execution will validate that there is enough reserve balance to pay for the intrinsic gas
- execution will still charge the carriage cost if the balance is insufficient to pay for the value transfer + gas limit

### CODE REVIEW

File: libs/execution/src/monad/chain/monad_chain.cpp [hunk 1]
```diff
@@ -1,10 +1,46 @@
...
+#include <monad/state3/state.hpp>
 
 MONAD_NAMESPACE_BEGIN
 
 using BOOST_OUTCOME_V2_NAMESPACE::success;
 
+namespace
+{
+    // TODO
+    constexpr uint256_t DEFAULT_MAX_RESERVE = 200'000;
```
Comment (mikety) [inline libs/execution/src/monad/chain/monad_chain.cpp]: The more practical value for max_reserve to pass validation is E17 - this is the number used in testing on devnet1.

File: libs/execution/src/monad/chain/monad_chain.cpp [hunk 2]
```diff
@@ -1,10 +1,46 @@
...
+
+    uint256_t effective_reserve(
+        IntrinsicGasBuffer const &buf, uint256_t const &max_reserve,
+        Address const &addr, uint256_t const &balance)
+    {
+        if (balance < max_reserve) {
+            return balance;
```
Comment (mikety) [inline libs/execution/src/monad/chain/monad_chain.cpp:21]: Can you please clarify why not balance - gas ?

File: libs/execution/src/monad/execution/intrinsic_gas_buffer.hpp [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+#pragma GCC diagnostic pop
+
+#include <cstdint>
+
+MONAD_NAMESPACE_BEGIN
+
+class IntrinsicGasBuffer
+{
+public:
+    static constexpr auto N = 4;
```
Comment (mikety) [inline libs/execution/src/monad/execution/intrinsic_gas_buffer.hpp]: Can you please clarify why N = 4? I assume this is the degree of parallelism - i.e number of blocks being processed in parallel.

File: libs/execution/src/monad/chain/monad_chain.cpp [hunk 4]
```diff
@@ -20,4 +56,60 @@ bool MonadChain::validate_root(
...
+    State &state, Transaction const &tx, Address const &sender,
+    std::optional<Account> const &acct)
+{
+    // Note: Intrinsic gas charged from the reserve balance
+    uint512_t const v0 =
+        tx.value +
+        max_gas_cost(tx.gas_limit - intrinsic_gas(rev, tx), tx.max_fee_per_gas);
+    auto const balance = acct.has_value() ? acct->balance : uint256_t{0};
+    if (effective_balance(DEFAULT_MAX_RESERVE, balance) < v0) {
+        return evmc::Result{EVMC_REJECTED};
```
Comment (mikety) [inline libs/execution/src/monad/chain/monad_chain.cpp:89]: Will this error be propagated to the receipt?

File: libs/execution/src/monad/chain/monad_chain.cpp [hunk 5]
```diff
@@ -20,4 +56,60 @@ bool MonadChain::validate_root(
...
+    auto res = ::monad::validate_transaction(tx, acc);
+    if (res.has_error() &&
+        res.error() != TransactionError::InsufficientBalance) {
+        return res;
+    }
+
+    uint256_t const balance = acc.has_value() ? acc->balance : uint256_t{0};
+    if (max_gas_cost(intrinsic_gas(rev, tx), tx.max_fee_per_gas) >
+        effective_reserve(buf_, DEFAULT_MAX_RESERVE, sender, balance)) {
+        return TransactionError::InsufficientReserveBalance;
```
Comment (mikety) [inline libs/execution/src/monad/chain/monad_chain.cpp:72]: Will this error be propagated to the receipt?

File: libs/execution/src/monad/chain/monad_chain.cpp (hunk 4 above)
Comment (zander-xyz) [reply libs/execution/src/monad/chain/monad_chain.cpp:89]: yes

File: libs/execution/src/monad/chain/monad_chain.cpp (hunk 5 above)
Comment (zander-xyz) [reply libs/execution/src/monad/chain/monad_chain.cpp:72]: no

File: libs/execution/src/monad/execution/intrinsic_gas_buffer.hpp (hunk 3 above)
Comment (zander-xyz) [reply libs/execution/src/monad/execution/intrinsic_gas_buffer.hpp]: it should be set to whatever the consensus delay is

File: libs/execution/src/monad/chain/monad_chain.cpp (hunk 2 above)
Comment (zander-xyz) [reply libs/execution/src/monad/chain/monad_chain.cpp:21]: that would be double charging

File: libs/execution/src/monad/execution/intrinsic_gas_buffer.hpp (hunk 3 above)
Comment (mikety) [reply libs/execution/src/monad/execution/intrinsic_gas_buffer.hpp]: in consensus execution_delay setting is 10. 
Please update. Nit: please add comment that N must match the consensus execution_delay .

File: libs/execution/src/monad/chain/monad_chain.cpp (hunk 1 above)
Comment (mikety) [reply libs/execution/src/monad/chain/monad_chain.cpp]: in consensus max_reserve_balance is 100_000_000_000_000_000,
Please update. Nit: please add comment that DEFAULT_MAX_RESERVE must match the consensus max_reserve_balance.

File: libs/execution/src/monad/chain/chain.hpp [hunk 6]
```diff
@@ -1,16 +1,23 @@
...
 
 #include <evmc/evmc.h>
+#include <optional>
 
 MONAD_NAMESPACE_BEGIN
 
+struct Account;
+class BlockHashBuffer;
 struct BlockHeader;
+class BlockState;
```
Comment (tongzhi1998) [inline libs/execution/src/monad/chain/chain.hpp]: We probably don't need BlockState.

Comment (jhunsaker) [reply libs/execution/src/monad/chain/chain.hpp]: removed

**Pull request closed without merging**